### ISSUE TITLE
Installation headers not found.

### ISSUE THREAD

**Issue by resyfer** [pgexporter/pgexporter_ext#7]
Issue: Can't find "postgres.h" in "lib.c"

Steps:
1) Followed the instructions of installation and installed `postgresql-devel` (rest were up to date)
2) Followed build instructions.

Given below is an asciinema recording: [![asciicast](https://asciinema.org/a/568146.svg)](https://asciinema.org/a/568146)

Things tried:
1) I tried to install the `postgresql-server-devel` package as that is supposed to have the `postgres.h` header file, but for some reason there is a conflict with `libpq` (installed from `postgresql-devel` as in the instructions)
2) I tried to copy the `postgres.h` header from postgres's source code, but it would require a lot of those headers being added to the code-base, so abandoned it.

**Comment by jesperpedersen** [comment]
What distribution is this ? You can see the necessary headers in `lib.c`

Seems strange there is a conflict between libpq and postgresql-server-devel...

**Comment by resyfer** [comment]
My Version information:

```sh
$ neofetch
             .',;::::;,'.                resyfer@fedora 
         .';:cccccccccccc:;,.            -------------- 
      .;cccccccccccccccccccccc;.         OS: Fedora Linux 37 (Workstation Edition) x86_64 
    .:cccccccccccccccccccccccccc:.       Host: Inspiron 3501 
  .;ccccccccccccc;.:dddl:.;ccccccc;.     Kernel: 6.1.18-200.fc37.x86_64 
 .:ccccccccccccc;OWMKOOXMWd;ccccccc:.    Uptime: 7 mins 
.:ccccccccccccc;KMMc;cc;xMMc:ccccccc:.   Packages: 2356 (rpm), 12 (flatpak) 
,cccccccccccccc;MMM.;cc;;WW::cccccccc,   Shell: bash 5.2.15 
:cccccccccccccc;MMM.;cccccccccccccccc:   Resolution: 1920x1080 
:ccccccc;oxOOOo;MMM0OOk.;cccccccccccc:   DE: GNOME 43.3 
cccccc:0MMKxdd:;MMMkddc.;cccccccccccc;   WM: Mutter 
ccccc:XM0';cccc;MMM.;cccccccccccccccc'   WM Theme: Adwaita 
ccccc;MMo;ccccc;MMW.;ccccccccccccccc;    Theme: Orchis-Dark [GTK2/3] 
ccccc;0MNc.ccc.xMMd:ccccccccccccccc;     Icons: Sevi-dark [GTK2/3] 
cccccc;dNMWXXXWM0::cccccccccccccc:,      Terminal: gnome-terminal 
cccccccc;.:odl:.;cccccccccccccc:,.       CPU: 11th Gen Intel i5-1135G7 (8) @ 4.200GHz 
:cccccccccccccccccccccccccccc:'.         GPU: Intel TigerLake-LP GT2 [Iris Xe Graphics] 
.:cccccccccccccccccccccc:;,..            Memory: 4404MiB / 7655MiB 
  '::cccccccccccccc::;,.
```
```sh
$ dnf list installed | grep libpq
libpq.x86_64                                         14.3-2.fc37                         @fedora                                         
libpq-devel.x86_64                                   14.3-2.fc37                         @fedora
```
```sh
$ uname -a
Linux fedora 6.1.18-200.fc37.x86_64 #1 SMP PREEMPT_DYNAMIC Sat Mar 11 16:09:14 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

Version of postgresql-server-devel that was about to be installed:
```sh
$ dnf list | grep postgresql-server-devel
postgresql-server-devel.i686                                                             14.3-8.fc37                                                      fedora                                          
postgresql-server-devel.x86_64                                                           14.3-8.fc37                                                      fedora
```

I also had tried:
```sh
$ pg_config --includedir-server
/usr/include/pgsql/server
```
But that file didn't exist. The only folder inside `/usr/include/pgsql` was `internal`.

---------------------------
PS:

I could finally get it to work by removing both `libpq` and `libpq-devel` and instead installing `postgresql-server-devel`. After that compiling and installing went without a hitch:

```sh
$ psql -c "SELECT * FROM pg_available_extensions WHERE name='pgexporter_ext';" postgres
      name      | default_version | installed_version |                comment                 
----------------+-----------------+-------------------+----------------------------------------
 pgexporter_ext | 0.3.0           |                   | pgexporter extension for extra metrics
(1 row)
```

Should I close this issue? Or make a small pr updating [postgresql-devel to postgresql-server-devel](https://github.com/pgexporter/pgexporter_ext/blob/main/README.md?plain=1#L34)?

**Comment by jesperpedersen** [comment]
Can you try with the RPMs from https://yum.postgresql.org/ ?

**Comment by resyfer** [comment]
Yes it worked.

Steps:
1) Stopped the postgresql service by `sudo systemctl stop postgresql`
2) Removed `postgresql-server-devel` by `sudo dnf remove postgresql-server-devel`
3) Tried the following (got it from the [form on Yum HowTo page](https://www.postgresql.org/download/linux/redhat/)) (added the -E beside sudo to provide my http_proxy env vars to dnf):
```sh
sudo -E dnf install -y https://download.postgresql.org/pub/repos/yum/reporpms/F-37-x86_64/pgdg-fedora-repo-latest.noarch.rpm

# Install PostgreSQL:
sudo -E dnf install -y postgresql15-server

# Optionally initialize the database and enable automatic start:
sudo /usr/pgsql-15/bin/postgresql-15-setup initdb
sudo systemctl enable postgresql-15
sudo systemctl start postgresql-15
```

and

```sh
$ sudo systemctl status postgresql-15
● postgresql-15.service - PostgreSQL 15 database server
     Loaded: loaded (/usr/lib/systemd/system/postgresql-15.service; enabled; preset: disabled)
     Active: active (running) since Sun 2023-03-19 09:14:43 IST; 1min 46s ago
       Docs: https://www.postgresql.org/docs/15/static/
    Process: 61543 ExecStartPre=/usr/pgsql-15/bin/postgresql-15-check-db-dir ${PGDATA} (code=exited, status=0/SUCCESS)
   Main PID: 61548 (postmaster)
      Tasks: 7 (limit: 9135)
     Memory: 23.4M
        CPU: 98ms
     CGroup: /system.slice/postgresql-15.service
             ├─61548 /usr/pgsql-15/bin/postmaster -D /var/lib/pgsql/15/data/
             ├─61549 "postgres: logger "
             ├─61550 "postgres: checkpointer "
             ├─61551 "postgres: background writer "
             ├─61553 "postgres: walwriter "
             ├─61554 "postgres: autovacuum launcher "
             └─61555 "postgres: logical replication launcher "

Mar 19 09:14:43 fedora systemd[1]: Starting postgresql-15.service - PostgreSQL 15 database server...
Mar 19 09:14:43 fedora postmaster[61548]: 2023-03-19 09:14:43.255 IST [61548] LOG:  redirecting log output to logging collector process
Mar 19 09:14:43 fedora postmaster[61548]: 2023-03-19 09:14:43.255 IST [61548] HINT:  Future log output will appear in directory "log".
Mar 19 09:14:43 fedora systemd[1]: Started postgresql-15.service - PostgreSQL 15 database server.
```

4) After adding `/usr/psql-15/bin` to PATH, it compiled and installed without any problems.

**Comment by resyfer** [comment]
TLDR: Works for yum packages `postgresql15-devel` and `postgresql-server-devel`.

Steps:
1) Stopped the postgresql service by `sudo systemctl stop postgresql`
2) Removed `postgresql-server-devel` by `sudo dnf remove postgresql-server-devel`
3) Tried the following (got it from the [form on Yum HowTo page](https://www.postgresql.org/download/linux/redhat/)) (added the -E beside sudo to provide my http_proxy env vars to dnf):
```sh
sudo -E dnf install -y https://download.postgresql.org/pub/repos/yum/reporpms/F-37-x86_64/pgdg-fedora-repo-latest.noarch.rpm

# Install PostgreSQL:
sudo -E dnf install -y postgresql15-server

# Optionally initialize the database and enable automatic start:
sudo /usr/pgsql-15/bin/postgresql-15-setup initdb
sudo systemctl enable postgresql-15
sudo systemctl start postgresql-15
```

and

```sh
$ sudo systemctl status postgresql-15
● postgresql-15.service - PostgreSQL 15 database server
     Loaded: loaded (/usr/lib/systemd/system/postgresql-15.service; enabled; preset: disabled)
     Active: active (running) since Sun 2023-03-19 09:14:43 IST; 1min 46s ago
       Docs: https://www.postgresql.org/docs/15/static/
    Process: 61543 ExecStartPre=/usr/pgsql-15/bin/postgresql-15-check-db-dir ${PGDATA} (code=exited, status=0/SUCCESS)
   Main PID: 61548 (postmaster)
      Tasks: 7 (limit: 9135)
     Memory: 23.4M
        CPU: 98ms
     CGroup: /system.slice/postgresql-15.service
             ├─61548 /usr/pgsql-15/bin/postmaster -D /var/lib/pgsql/15/data/
             ├─61549 "postgres: logger "
             ├─61550 "postgres: checkpointer "
             ├─61551 "postgres: background writer "
             ├─61553 "postgres: walwriter "
             ├─61554 "postgres: autovacuum launcher "
             └─61555 "postgres: logical replication launcher "

Mar 19 09:14:43 fedora systemd[1]: Starting postgresql-15.service - PostgreSQL 15 database server...
Mar 19 09:14:43 fedora postmaster[61548]: 2023-03-19 09:14:43.255 IST [61548] LOG:  redirecting log output to logging collector process
Mar 19 09:14:43 fedora postmaster[61548]: 2023-03-19 09:14:43.255 IST [61548] HINT:  Future log output will appear in directory "log".
Mar 19 09:14:43 fedora systemd[1]: Started postgresql-15.service - PostgreSQL 15 database server.
```

4) After adding `/usr/psql-15/bin` (for `pg_config`) to PATH, I tried to compile it but same problem that it can't find the header file.

5) Installed `postgresql-server-devel` using yum, and compiles like in dnf.

6) I also

**Comment by resyfer** [comment]
TLDR: Works for yum packages `postgresql15-devel` and `postgresql-server-devel`.

Steps:
1) Stopped the postgresql service by `sudo systemctl stop postgresql`
2) Removed `postgresql-server-devel` by `sudo dnf remove postgresql-server-devel`
3) Tried the following (got it from the [form on Yum HowTo page](https://www.postgresql.org/download/linux/redhat/)) (added the -E beside sudo to provide my http_proxy env vars to dnf):
```sh
sudo -E dnf install -y https://download.postgresql.org/pub/repos/yum/reporpms/F-37-x86_64/pgdg-fedora-repo-latest.noarch.rpm

# Install PostgreSQL:
sudo -E dnf install -y postgresql15-server

# Optionally initialize the database and enable automatic start:
sudo /usr/pgsql-15/bin/postgresql-15-setup initdb
sudo systemctl enable postgresql-15
sudo systemctl start postgresql-15
```

and

```sh
$ sudo systemctl status postgresql-15
● postgresql-15.service - PostgreSQL 15 database server
     Loaded: loaded (/usr/lib/systemd/system/postgresql-15.service; enabled; preset: disabled)
     Active: active (running) since Sun 2023-03-19 09:14:43 IST; 1min 46s ago
       Docs: https://www.postgresql.org/docs/15/static/
    Process: 61543 ExecStartPre=/usr/pgsql-15/bin/postgresql-15-check-db-dir ${PGDATA} (code=exited, status=0/SUCCESS)
   Main PID: 61548 (postmaster)
      Tasks: 7 (limit: 9135)
     Memory: 23.4M
        CPU: 98ms
     CGroup: /system.slice/postgresql-15.service
             ├─61548 /usr/pgsql-15/bin/postmaster -D /var/lib/pgsql/15/data/
             ├─61549 "postgres: logger "
             ├─61550 "postgres: checkpointer "
             ├─61551 "postgres: background writer "
             ├─61553 "postgres: walwriter "
             ├─61554 "postgres: autovacuum launcher "
             └─61555 "postgres: logical replication launcher "

Mar 19 09:14:43 fedora systemd[1]: Starting postgresql-15.service - PostgreSQL 15 database server...
Mar 19 09:14:43 fedora postmaster[61548]: 2023-03-19 09:14:43.255 IST [61548] LOG:  redirecting log output to logging collector process
Mar 19 09:14:43 fedora postmaster[61548]: 2023-03-19 09:14:43.255 IST [61548] HINT:  Future log output will appear in directory "log".
Mar 19 09:14:43 fedora systemd[1]: Started postgresql-15.service - PostgreSQL 15 database server.
```

4) After adding `/usr/psql-15/bin` (for `pg_config`) to PATH, I tried to compile it but same problem that it can't find the header file.

5) Installed `postgresql-server-devel` using yum, and **it compiles well**. I then removed the package.

6) Then I installed the `postgresql15-devel` using yum, **which works as well** (there was an implicit definition of `getnameinfo` in lib.c, which upon addition of `netdb.h` allowed compiling)

**Comment by Aditya0049** [comment]
Stop the PostgreSQL service by running the following command:

vbnet
`sudo systemctl stop postgresql`

Remove the existing "postgresql-server-devel" package by running the following command:

csharp
`sudo dnf remove postgresql-server-devel`

Install the latest PostgreSQL repository package using the following command:

ruby
`sudo -E dnf install -y https://download.postgresql.org/pub/repos/yum/reporpms/F-37-x86_64/pgdg-fedora-repo-latest.noarch.rpm`

Install the "postgresql15-server" package using the following command:

`sudo -E dnf install -y postgresql15-server`

Initialize the database and start the PostgreSQL service using the following commands:

bash
```
sudo /usr/pgsql-15/bin/postgresql-15-setup initdb
sudo systemctl enable postgresql-15
sudo systemctl start postgresql-15
```

Add the path "/usr/pgsql-15/bin" to the PATH environment variable to access the "pg_config" utility.

Install the "postgresql15-devel" package using the following command:

`sudo -E dnf install -y postgresql15-devel`
Compile the program again. This time, it should be able to find the "postgres.h" header file.

**Comment by resyfer** [comment]
@Aditya0049 Thank you for the reply. It was partly my fault for not making it clear, but I've already done those steps. However, I do not understand why each bash script has a different language written above it.

On that note @jesperpedersen, should I make a PR that updates the dependencies to either `postgresql-server-devel` from DNF or the YUM repositories of `postgresql15-server` and `postgresql15-devel`? If yes, should I add both or any specific one or use a specific version of PostgreSQL for YUM (like 10 instead of 15)?

**Comment by jesperpedersen** [comment]
Packages will be named differently depending on their repository.

I think it is enough to high-light that the dependencies are from yum.postgresql.org - and not from the distribution repository.
